src/v/raft/state_machine.cc

```diff
@@ -79,6 +79,14 @@ state_machine::batch_applicator::operator()(model::record_batch batch) {
           ss::stop_iteration::yes);
     }
 
+    const auto committed_offset = _machine->_raft->committed_offset();
+    vassert(
+      batch.last_offset() <= committed_offset,
+      "Can not apply not committed batches to stm. Applied batch header: {}, "
+      "raft committed offset: {}",
+      batch.header(),
+      committed_offset);
```

Comment:
if this assertion fails, will we be able to know which raft group and which state machine had the failure? i could see that being useful because presumably we'd want to go back through logs looking at that particular raft group?